src/request.rs

```diff
@@ -1,220 +1,390 @@
 use crate::errors::{Error, MeilisearchError};
+use async_trait::async_trait;
 use log::{error, trace, warn};
 use serde::{de::DeserializeOwned, Serialize};
 use serde_json::{from_str, to_string};
 
 #[derive(Debug)]
-pub(crate) enum Method<Q, B> {
+pub enum Method<Q, B> {
     Get { query: Q },
     Post { query: Q, body: B },
     Patch { query: Q, body: B },
     Put { query: Q, body: B },
     Delete { query: Q },
 }
 
-#[cfg(not(target_arch = "wasm32"))]
-pub fn add_query_parameters<Query: Serialize>(url: &str, query: &Query) -> Result<String, Error> {
-    let query = yaup::to_string(query)?;
+#[async_trait(?Send)]
+pub trait HttpClient: Clone + Serialize + Send + Sync {
+    #[cfg(not(target_arch = "wasm32"))]
+    async fn request<Query, Body, Output>(
+        self,
+        url: &str,
+        apikey: Option<&str>,
+        method: Method<Query, Body>,
+        expected_status_code: u16,
+    ) -> Result<Output, Error>
+    where
+        Query: Serialize + Send + Sync,
+        Body: Serialize + Send + Sync,
+        Output: DeserializeOwned + 'static + Send;
+
+    #[cfg(target_arch = "wasm32")]
+    async fn request<Query, Body, Output>(
+        self,
+        url: &str,
+        apikey: Option<&str>,
+        method: Method<Query, Body>,
+        expected_status_code: u16,
+    ) -> Result<Output, Error>
+    where
+        Query: Serialize + Send + Sync,
+        Body: Serialize + Send + Sync,
+        Output: DeserializeOwned + 'static + Send;
```

Comment:
```suggestion
    async fn request<Query, Body, Output>(
        self,
        url: &str,
        apikey: Option<&str>,
        method: Method<Query, Body>,
        expected_status_code: u16,
    ) -> Result<Output, Error>
    where
        Query: Serialize + Send + Sync,
        Body: Serialize + Send + Sync,
        Output: DeserializeOwned + 'static + Send;
```

Huum, I’m struggling to see a difference between these two 🤔 
Shouldn’t we be able to simply define it once without the wasm/no-wasm attribute?

Reply:
You are right!
I didn't bother to separate them 😂
fixed
https://github.com/meilisearch/meilisearch-rust/pull/459/commits/2a1cc7a8e6b088c15c7c34ac3fe98246c1b96947